### PR TITLE
Type annotations for sqlalchemy.orm.dynamic

### PR DESCRIPTION
**Pull request by g-kisenkov**
### Description

An attempt to annotate lib/sqlalchemy/orm/dynamic.py with type hints (issue https://github.com/sqlalchemy/sqlalchemy/issues/6810)

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: lib/sqlalchemy/orm/dynamic.py [hunk 1]
```diff
@@ -21,35 +20,56 @@
...
 from .session import object_session
+from .util import AliasedClass
 from .writeonly import AbstractCollectionWriter
 from .writeonly import WriteOnlyAttributeImpl
+from .writeonly import WriteOnlyCollection
 from .writeonly import WriteOnlyHistory
 from .writeonly import WriteOnlyLoader
 from .. import util
 from ..engine import result
+from ..event import _Dispatch
```
Comment (g-kisenkov) [inline lib/sqlalchemy/orm/dynamic.py:53]: @mlatysh 

All internal modules must be imported relatively and put under an `if TYPE_CHECKING` branch if not used at runtime.

Example: https://github.com/GodelTech/sqlalchemy/pull/3/files#diff-a178567dd254a0b999fc247ba30743808c234d94f43b7cc88d878be1ec31fe55R40-R55

File: lib/sqlalchemy/orm/dynamic.py [hunk 2]
```diff
@@ -67,27 +87,32 @@ class DynamicAttributeImpl(WriteOnlyAttributeImpl):
...
-        query_class=None,
-        **kw,
-    ):
+        class_: Union[Type[_T], AliasedClass[_T]],
+        key: str,
+        typecallable: type,
+        dispatch: _Dispatch[QueryableAttribute[_T]],
+        target_mapper: Mapper[_T],
+        order_by: Tuple[()],
+        # Tuple[Union[bool, UnaryExpression[_T], Column[_T]]]
```
Comment (g-kisenkov) [inline lib/sqlalchemy/orm/dynamic.py:96]: Is it a leftover from running PyAnnotate with Python 2 support?

File: lib/sqlalchemy/orm/dynamic.py (hunk 1 above)
Comment (mlatysh) [reply lib/sqlalchemy/orm/dynamic.py]: @g-kisenkov @dzmitar Refactored as mentioned, made new imports under `if TYPE_CHECKING` condition and made all of them relative

**Comment by g-kisenkov** [comment]
@mlatysh 

The PR is cleared on my end, if @dzmitar has also finished reviewing the changes you can open a PR into the base SQLA repo.

**Review by mlatysh — APPROVED** [review]

**Pull request closed without merging**